src/Panel/Controller/Changes.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+namespace Kirby\Panel\Controller;
+
+use Kirby\Cms\ModelWithContent;
+use Kirby\Content\VersionId;
+use Kirby\Exception\Exception;
+use Kirby\Form\Form;
+
+/**
+ * The Changes controller takes care of the request logic
+ * to save, discard and publish changes, as well as unlocking.
+ *
+ * @package   Kirby Panel
+ * @author    Bastian Allgeier <bastian@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://getkirby.com/license
+ */
+class Changes
+{
+	/**
+	 * Discards unpublished changes by deleting the version
+	 */
+	public static function discard(ModelWithContent $model): array
+	{
+		$model->version(VersionId::changes())->delete();
+
+		return [
+			'status' => 'ok'
+		];
+	}
+
+	/**
+	 * Saves the lastest state of changes first and then publishs them
+	 */
+	public static function publish(ModelWithContent $model, array $input): array
+	{
+		// save the given changes first
+		static::save(
+			model: $model,
+			input: $input
+		);
+
+		// get the changes version
+		$changes = $model->version(VersionId::changes());
+
+		// publish the changes
+		$changes->publish(
+			language: 'current'
+		);
+
+		return [
+			'status' => 'ok'
+		];
+	}
+
+	/**
+	 * Saves form input in a new or existing `changes` version
+	 */
+	public static function save(ModelWithContent $model, array $input): array
+	{
+		// we need to run the input through the form
+		// class to get a set of storable field values
+		// that we can send to the content storage handler
+		$form = Form::for($model, [
+			'ignoreDisabled' => true,
+			'input'          => $input,
+		]);
```

Comment:
`Question:` Isn't this something the `Version` class needs to handle as it would also apply to programmatic updates not through this Panel controller class?

Reply:
Yes, we are getting to the point where this needs to be in the Version class, but I honestly want to spend more time on it and it should be in a separate PR in my opinion.